core/src/main/java/com/linecorp/armeria/common/stream/ConcatPublisherStreamMessage.java

```diff
@@ -0,0 +1,313 @@
+        private volatile int cancelled;
+
+        InnerSubscriber(Subscriber<? super T> downstream, SubscriptionOption[] options,
+                        ConcatPublisherStreamMessage<T> publisher) {
+            this.downstream = downstream;
+            this.options = options;
+            this.publisher = publisher;
+        }
+
+        /**
+         * Sets the {@link OuterSubscriber}.
+         * This should be set before subscribing to upstream.
+         */
+        void setOuterSubscriber(OuterSubscriber<T> outerSubscriber) {
+            this.outerSubscriber = outerSubscriber;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (isCancelled()) {
+                subscription.cancel();
+                return;
+            }
+            // Reset 'completed' to subscribe to new Publisher
+            completed = false;
+            setUpstreamSubscription(subscription);
+            outerSubscriber.inInnerOnSubscribe = false;
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            if (isCancelled()) {
+                return;
+            }
+            downstream.onNext(item);
+            produced(1);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+            if (isCancelled()) {
+                return;
+            }
+            downstream.onError(cause);
+            publisher.abort(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            if (isCancelled()) {
+                return;
+            }
+            completed = true;
+            if (outerSubscriber.completed) {
+                downstream.onComplete();
+            } else {
+                outerSubscriber.nextSource();
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                downstream.onError(new IllegalArgumentException(
+                        "Rule §3.9 violated: non-positive requests are forbidden"));
+                return;
+            }
+            if (isCancelled()) {
+                return;
+            }
+            super.request(n);
+        }
+
+        @Override
+        public void cancel() {
+            if (cancelledUpdater.compareAndSet(this, 0, 1)) {
+                if (outerSubscriber != null) {
+                    final Subscription upstream = outerSubscriber.upstream;
+                    if (upstream != null) {
+                        upstream.cancel();
+                    }
+                }
+                super.cancel();
+                final CancelledSubscriptionException cause = CancelledSubscriptionException.get();
+                publisher.abort(cause);
+                if (!completed && containsNotifyCancellation(options)) {
```

Comment:
Shouldn't we also check if `onError` from the publisher is called or not?

Reply:
Good suggestion. Added 'error' flag to avoid this race condition.